packages/go_router/lib/src/misc/stateful_navigation_shell.dart

```diff
@@ -0,0 +1,397 @@
+    return routeState != oldWidget.routeState;
+  }
+}
+
+/// Widget that manages and maintains the state of a [StatefulShellRoute],
+/// including the [Navigator]s of the configured route branches.
+///
+/// This widget acts as a wrapper around the builder function specified for the
+/// associated StatefulShellRoute, and exposes the state (represented by
+/// [StatefulShellRouteState]) to its child widgets with the help of the
+/// InheritedWidget [InheritedStatefulNavigationShell]. The state for each route
+/// branch is represented by [StatefulShellBranchState] and can be accessed via the
+/// StatefulShellRouteState.
+///
+/// By default, this widget creates a container for the branch route Navigators,
+/// provided as the child argument to the builder of the StatefulShellRoute.
+/// However, implementors can choose to disregard this and use an alternate
+/// container around the branch navigators
+/// (see [StatefulShellRouteState.children]) instead.
+class StatefulNavigationShell extends StatefulWidget {
+  /// Constructs an [StatefulNavigationShell].
+  const StatefulNavigationShell({
+    required this.shellRoute,
+    required this.navigatorBuilder,
+    required this.shellBodyWidgetBuilder,
+    super.key,
+  });
+
+  /// The associated [StatefulShellRoute]
+  final StatefulShellRoute shellRoute;
+
+  /// The shell navigator builder.
+  final ShellNavigatorBuilder navigatorBuilder;
+
+  /// The shell body widget builder.
+  final ShellBodyWidgetBuilder shellBodyWidgetBuilder;
+
+  @override
+  State<StatefulWidget> createState() => StatefulNavigationShellState();
+}
+
+/// State for StatefulNavigationShell.
+class StatefulNavigationShellState extends State<StatefulNavigationShell> {
+  final Map<Key, Widget> _navigatorCache = <Key, Widget>{};
+
+  late StatefulShellRouteState _routeState;
+
+  List<StatefulShellBranch> get _branches => widget.shellRoute.branches;
+
+  GoRouterState get _currentGoRouterState => widget.navigatorBuilder.state;
+  GlobalKey<NavigatorState> get _currentNavigatorKey =>
+      widget.navigatorBuilder.navigatorKeyForCurrentRoute;
+
+  Widget? _navigatorForBranch(StatefulShellBranch branch) {
+    return _navigatorCache[branch.navigatorKey];
+  }
+
+  void _setNavigatorForBranch(StatefulShellBranch branch, Widget? navigator) {
+    navigator != null
+        ? _navigatorCache[branch.navigatorKey] = navigator
+        : _navigatorCache.remove(branch.navigatorKey);
+  }
+
+  int _findCurrentIndex() {
+    final int index = _branches.indexWhere(
+        (StatefulShellBranch e) => e.navigatorKey == _currentNavigatorKey);
+    assert(index >= 0);
+    return index;
+  }
+
+  void _switchActiveBranch(
+      StatefulShellBranchState branchState, bool resetLocation) {
+    final GoRouter goRouter = GoRouter.of(context);
+    final GoRouterState? routeState = branchState.routeState;
+    if (routeState != null && !resetLocation) {
+      goRouter.goState(routeState, context).onError(
```

Comment:
This should probably go directly through RouterDelegate without going through the goRouter. We can use setNewRoutePath directly.

Reply:
Yes, did that before, but created the goState method in GoRouter as an alternative to StatefulShellRouteState.goBranch (which could then potentially be removed). However a better replacement could perhaps instead be a goBranch method on GoRouter, that accepts a StatefulShellBranchState as argument. What do you think?

Reply:
Replacing goState with goBranch would also mean I could revert the changes to GoRouterState, and return to storing a RouteMatchList/UnmodifiableRouteMatchList in StatefulShellBranchState.

Reply:
If developer needs to pass in branchState, we may as well ask them to get the StatefulShellRouteState and do gobranch from it directly.

What I am trying to do is to reduce # of apis that does the same thing if they are fairly easy to do